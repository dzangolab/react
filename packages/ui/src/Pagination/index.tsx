import React, { useState, useEffect, useCallback } from "react";

import { Button, DebouncedInput } from "..";

export interface PaginationProperties {
  currentPage: number;
  totalItems: number;
  onPageChange: (page: number) => void;
  onItemsPerPageChange?: (itemsPerPage: number) => void;
  itemsPerPageOptions?: number[];
  itemsPerPageControlLabel?: string;
  inputDebounceTime?: number;
  defaultItemsPerPage?: number;
  showFirstLastButtons?: boolean;
  showPreviousNextButtons?: boolean;
  showPageButtons?: boolean;
  className?: string;
  pageInputLabel?: string;
  showItemsPerPageControl?: boolean;
  showPageInput?: boolean;
}

export const Pagination: React.FC<PaginationProperties> = ({
  currentPage,
  totalItems,
  onPageChange,
  onItemsPerPageChange,
  itemsPerPageOptions = [5, 10, 20, 30],
  itemsPerPageControlLabel = "Items per page",
  inputDebounceTime,
  pageInputLabel = "Go to page:",
  defaultItemsPerPage = itemsPerPageOptions[0],
  showFirstLastButtons = true,
  showPageButtons = false,
  showPreviousNextButtons = true,
  className,
  showItemsPerPageControl = true,
  showPageInput = true,
}) => {
  const [itemsPerPage, setItemsPerPage] = useState<number>(defaultItemsPerPage);

  useEffect(() => {
    setItemsPerPage(defaultItemsPerPage);
    onPageChange(0);
  }, [defaultItemsPerPage]);

  const lastPage = Math.ceil(totalItems / itemsPerPage);
  const pages = Array.from({ length: lastPage }, (_, index) => index + 1);

  const handleItemsPerPageChange = (
    event: React.ChangeEvent<HTMLSelectElement>,
  ) => {
    const newItemsPerPage = parseInt(event.target.value, 10);
    setItemsPerPage(newItemsPerPage);
    onItemsPerPageChange && onItemsPerPageChange(newItemsPerPage);
  };

  const handlePageInputChange = useCallback(
    (value: string | number | readonly string[]) => {
      const newPage = parseInt(value.toString(), 10) - 1;

      if (!isNaN(newPage) && newPage >= 0 && newPage < lastPage) {
        onPageChange(newPage);
      }
    },
    [lastPage, onPageChange],
  );

  const pageStatics = <span> {`${currentPage + 1} / ${lastPage}`}</span>;

  return (
    <div className={`pagination ${className || ""}`}>
      {showItemsPerPageControl && (
        <div className="items-per-page-control">
          <span>{itemsPerPageControlLabel}</span>
          <select value={itemsPerPage} onChange={handleItemsPerPageChange}>
            {itemsPerPageOptions.map((option) => (
              <option key={option} value={option}>
                {option}
              </option>
            ))}
          </select>
        </div>
      )}

      {lastPage > 1 && showPageInput && (
        <div className="page-input-control">
          <span>{pageInputLabel}</span>
          <DebouncedInput
            type="number"
            debounceTime={inputDebounceTime}
            onInputChange={handlePageInputChange}
          />
        </div>
      )}
      {lastPage > 1 ? (
        <div className="pagination-buttons-wrapper">
          {showFirstLastButtons || showPreviousNextButtons ? (
            <div>
              {showFirstLastButtons && (
                <Button
                  className="first-page"
                  onClick={() => onPageChange(0)}
                  disabled={!(currentPage > 0)}
                  iconLeft={<i className="pi pi-angle-double-left" />}
                />
              )}

              {showPreviousNextButtons && (
                <Button
                  className="previous-page"
                  onClick={() => onPageChange(currentPage - 1)}
                  disabled={!(currentPage > 0)}
                  iconLeft={<i className="pi pi-angle-left" />}
                />
              )}
            </div>
          ) : null}

          <div>
            {showPageButtons
              ? pages.map((page) => (
                  <Button
                    key={page}
                    className={`page-button ${
                      page === currentPage + 1 ? "active" : ""
                    }`}
                    onClick={() => onPageChange(page - 1)}
                    label={`${page}`}
                  />
                ))
              : pageStatics}
          </div>

          {showFirstLastButtons || showPreviousNextButtons ? (
            <div>
              {showPreviousNextButtons && (
                <Button
                  className="next-page"
                  onClick={() => onPageChange(currentPage + 1)}
                  disabled={!(currentPage < lastPage - 1)}
                  iconLeft={<i className="pi pi-angle-right" />}
                />
              )}

              {showFirstLastButtons && (
                <Button
                  className="last-page"
                  onClick={() => onPageChange(lastPage - 1)}
                  disabled={!(currentPage < lastPage - 1)}
                  iconLeft={<i className="pi pi-angle-double-right" />}
                />
              )}
            </div>
          ) : null}
        </div>
      ) : (
        pageStatics
      )}
    </div>
  );
};
