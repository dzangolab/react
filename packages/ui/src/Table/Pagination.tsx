import { useContext, useId } from "react";

import { TableContext } from "./TableProvider";

const Pagination = () => {
  const id = useId();
  const { isLoading, paginationIcons, table } = useContext(TableContext);

  const config = [
    {
      icon: paginationIcons?.start,
      isDisabled: !table?.getCanPreviousPage(),
      onClick: () => table?.setPageIndex(0),
    },
    {
      icon: paginationIcons?.previous,
      isDisabled: !table?.getCanPreviousPage(),
      onClick: () => table?.previousPage(),
    },
    {
      icon: paginationIcons?.next,
      isDisabled: !table?.getCanNextPage(),
      onClick: () => table?.nextPage(),
    },
    {
      icon: paginationIcons?.end,
      isDisabled: !table?.getCanNextPage(),
      onClick: () => table?.setPageIndex(table.getPageCount() - 1),
    },
  ];

  return (
    <div>
      {config.map(({ icon, isDisabled, onClick }, index) => (
        <button
          key={id + index}
          onClick={onClick}
          disabled={isDisabled || isLoading}
        >
          <img src={icon} />
        </button>
      ))}
    </div>
  );
};

export default Pagination;
