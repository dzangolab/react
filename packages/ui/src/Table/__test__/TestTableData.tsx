const data = [
  { id: 1, name: "jone" },
  { id: 2, name: "abhi" },
  { id: 3, name: "mike" },
  { id: 4, name: "tony" },
];

const biggerData = [
  { id: 1, name: "jone" },
  { id: 2, name: "abhi" },
  { id: 3, name: "mike" },
  { id: 4, name: "tony" },
  { id: 5, name: "tony" },
  { id: 6, name: "tony" },
  { id: 7, name: "tony" },
  { id: 8, name: "tony" },
  { id: 9, name: "tony" },
  { id: 10, name: "tony" },
  { id: 11, name: "tony" },
  { id: 12, name: "tony" },
  { id: 13, name: "tony" },
  { id: 14, name: "tony" },
  { id: 15, name: "tony" },
];

const fetcher = () => {
  return data;
};

const columns = [
  {
    accessorKey: "id",
    header: () => <span>ID</span>,
  },
  {
    accessorKey: "name",
    header: () => <span>Name</span>,
  },
];

export { columns, data, fetcher, biggerData };
