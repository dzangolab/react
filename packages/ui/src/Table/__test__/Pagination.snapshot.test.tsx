import React from "react";
import renderer, { create } from "react-test-renderer";
import { expect, test } from "vitest";

import { columns, data, fetcher } from "./TestTableData";
import Pagination from "../common/Pagination";
import { Table } from "../index";

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = create(
    <Table
      columns={columns}
      data={data}
      fetcher={fetcher}
      totalItems={data.length}
    >
      <Pagination />
    </Table>
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
