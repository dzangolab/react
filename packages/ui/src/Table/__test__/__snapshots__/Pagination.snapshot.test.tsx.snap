// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<div>
  <button
    disabled={true}
    onClick={[Function]}
  >
    <img
      src="/src/assets/images/double-chevron-left.svg"
    />
  </button>
  <button
    disabled={true}
    onClick={[Function]}
  >
    <img
      src="/src/assets/images/chevron-left.svg"
    />
  </button>
  <button
    onClick={[Function]}
  >
    <img
      src="/src/assets/images/chevron-right.svg"
    />
  </button>
  <button
    onClick={[Function]}
  >
    <img
      src="/src/assets/images/double-chevron-right.svg"
    />
  </button>
</div>
`;
